optaplanner-examples/src/main/java/org/optaplanner/examples/coachshuttlegathering/domain/solver/BusPassengerCountTotalUpdatingVariableListener.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright 2020 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.optaplanner.examples.coachshuttlegathering.domain.solver;
+
+import org.optaplanner.core.api.score.director.ScoreDirector;
+import org.optaplanner.core.impl.domain.variable.listener.VariableListener;
+import org.optaplanner.examples.coachshuttlegathering.domain.Bus;
+import org.optaplanner.examples.coachshuttlegathering.domain.BusStop;
+import org.optaplanner.examples.coachshuttlegathering.domain.CoachShuttleGatheringSolution;
+
+public abstract class BusPassengerCountTotalUpdatingVariableListener implements VariableListener<Object> {
+
+    @Override
+    public void beforeEntityAdded(ScoreDirector scoreDirector, Object busStop) {
+        // Do nothing
+    }
+
+    @Override
+    public void afterEntityAdded(ScoreDirector scoreDirector, Object entity) {
+        if (entity instanceof BusStop) {
+            updateBusPassengerCount(scoreDirector, (BusStop) entity, true);
+        }
+    }
+
+    @Override
+    public void beforeVariableChanged(ScoreDirector scoreDirector, Object entity) {
+        if (entity instanceof BusStop) {
+            updateBusPassengerCount(scoreDirector, (BusStop) entity, false);
+        }
+    }
+
+    @Override
+    public void afterVariableChanged(ScoreDirector scoreDirector, Object entity) {
+        if (entity instanceof BusStop) {
+            updateBusPassengerCount(scoreDirector, (BusStop) entity, true);
+        }
+    }
+
+    @Override
+    public void beforeEntityRemoved(ScoreDirector scoreDirector, Object entity) {
+        // Do nothing
+    }
+
+    @Override
+    public void afterEntityRemoved(ScoreDirector scoreDirector, Object entity) {
+        if (entity instanceof BusStop) {
+            updateBusPassengerCount(scoreDirector, (BusStop) entity, false);
+        }
+    }
+
+    private void updateBusPassengerCount(ScoreDirector<CoachShuttleGatheringSolution> scoreDirector, BusStop busStop,
+            boolean increase) {
+        Bus bus = busStop.getBus();
+        if (!isCorrectBusInstance(bus)) {
+            return;
+        }
+        int difference = increase ? busStop.getPassengerQuantity() : -busStop.getPassengerQuantity();
+        if (difference == 0) {
+            return;
+        }
```

Comment:
We can probably improve this:

    int quantity = busStop.getPassengerQuantity();
    if (quantity == 0) {
        return;
    }
    int difference = increase ? quantity : -quantity;

If this code is called often, this will bring a small perf improvement.

Reply:
As a side note, can a bus stop have 0 passengers? I don't think it can (and if it does, it an error; passengerQuantity is NOT a shadow variable or a planning variable)

Reply:
Good point. Maybe we should remove that check altogether.